Increase unittest `check_logcdf` coverage and fix issues with some distribution methods

ricardoV94 opened the pull request:
Changes:
This PR proposes two changes to `check_logcdf` to:
1. Test that values below or above finite domain edges are properly evaluated to `-inf` and `0`. 
2. Test that multiple values can be computed in one call, or if not, that a standard informative `TypeError` is returned.

This extended test revealed some issues with current implementations of the `logcdf` method, which are also fixed by this PR:
1. `Uniform`: Would incorrectly evaluate values above `upper` to `-inf`
2. `HalfNormal`: Would return `nan` for negative values. It was also unclear to me what the previous `logcdf` logic was doing, as it seems to me that `tt.lt(z, -1.0)` could never evaluate to `True` given valid positive values and `sigma`. As such, I simplified the code as well as correcting the invalid logcdf evaluation for negative values. Am I missing something? (pinging @domenzain)
3. `Gamma`, and `InverseGamma` were currently failing with invalid parameters due to #4340 and https://github.com/pymc-devs/Theano-PyMC/issues/224. I added a dirty hack to hide the issue for now. It can be simplified once those issues are solved.
4. `Beta` and `StudentT` logcdf now raises informative `TypeError` when asked to evaluate multiple values (#4342)

As suggested by @AlexAndorra in https://github.com/pymc-devs/pymc3/pull/4387#discussion_r549825529, the docstrings of the logcdf methods were updated to better reflect when multiple values can or cannot be evaluated.

It was necessary to change the test domains of the `Flat` and `HalfFlat` distributions to their actual domains.

This test is also relevant for the current PR #4387, were these same changes were helpful in debugging issues with the implementations of the `logcdf` methods. I thought it was best to have this as a separate PR as that one is quite large already. 

Finally, I think this does meaningfully increase the overhead of the unit tests since, at most, only three extra logp evaluations are being done for each distribution (one for each finite edge of a domain plus a (2,) array for multiple values).

ricardoV94 committed:
Change `check_logcdf` to test that values below or above domain are properly evaluated.
Fix issues with `Uniform`, `HalfNormal`, `Gamma`, and `InverseGamma` distributions.

ricardoV94 commented:
Failing test is irrelevant (flaky random MvNormal issues)

AlexAndorra reviewed (approved):
Looks all good, thanks @ricardoV94 ! I'll leave this open 24 hours before merging, in case someone wants to comment 😉

domenzain commented on pymc3/distributions/continuous.py:
```diff
@@ -2478,7 +2481,17 @@ def logcdf(self, value):
         beta = self.beta
-        return bound(tt.log(tt.gammainc(alpha, beta * value)), value >= 0, alpha > 0, beta > 0)
+        # To avoid issue with #4340
+        safe_alpha = tt.switch(tt.lt(alpha, 0), 0, alpha)
```
It is generally better practice to raise an Exception on invalid input (`ValueError`) than to silently change the input.

ricardoV94 replied:
I answered this in the main conversation.

domenzain commented on pymc3/distributions/continuous.py:
```diff
@@ -2478,7 +2481,17 @@ def logcdf(self, value):
         alpha = self.alpha
         beta = self.beta
-        return bound(tt.log(tt.gammainc(alpha, beta * value)), value >= 0, alpha > 0, beta > 0)
+        # To avoid issue with #4340
```
Better to say what the actual issue is rather than point elsewhere.
Better still to explain this in the commit message.

ricardoV94 replied:
Agree, will do.

domenzain commented on pymc3/distributions/continuous.py:
```diff
@@ -910,10 +913,10 @@ def logcdf(self, value):
         """
         sigma = self.sigma
         z = zvalue(value, mu=0, sigma=sigma)
-        return tt.switch(
```
The logic here was to have numerically stable `erfc` function.
Indeed not relevant for reasonable inputs for the `logcdf` of a `HalfNormal`, but maybe useful to abstract away into another function?

ricardoV94 replied:
Thanks for clarifying the purpose. Do you know of any places in the existing codebase that could benefit from that more stable `erfc`?  If you think it is useful for users directly as well, would you like to implement a PR for it?

ricardoV94 replied:
NVM: I think it already exists in `normal_lcdf` https://github.com/pymc-devs/pymc3/blob/b7daf266ef4247e33b32db069800180627bd731c/pymc3/distributions/dist_math.py#L128-L135

domenzain commented on pymc3/distributions/continuous.py:
```diff
@@ -2642,8 +2655,13 @@ def logcdf(self, value):
         """
         alpha = self.alpha
         beta = self.beta
+        # To avoid issue with #4340
```
Again better explain the actual issue here rather than point elsewhere. And better still to put this information right in the commit message.

domenzain commented on pymc3/distributions/continuous.py:
```diff
@@ -2642,8 +2655,13 @@ def logcdf(self, value):
         beta = self.beta
+        # To avoid issue with #4340
+        safe_alpha = tt.switch(tt.lt(alpha, 0), 0, alpha)
+        safe_beta = tt.switch(tt.lt(beta, 0), 0, beta)
```
Again, I would advise against silently modifying user input.
Better raise `ValueError` is the user is being unreasonable with a message explaining why.

domenzain reviewed (commented):
Hi @ricardoV94, 

I've left a couple of comments in a code review for your changes.
The extra hoops for the logic in the original implementation of the `logcdf` for the `HalfNormal` distribution were actually about the `erfc` function for input outside reasonable parameters for the distribution itself.

Best,

ricardoV94 commented:
> I've left a couple of comments in a code review for your changes.

Thank you for your review.

I disagree with the `ValueError` suggestion. This is an issue with the current implementation of the C implementation of the gamma functions in Theano. 

My solution does not give an unexpected behavior. It gives the same results one gets when inputting invalid parameters/ values in the `logcdf` method of other distributions (i.e. `-inf`), which is also the behavior of Scipy. It exists simply to avoid the debilitating C-assertion that exits the Python process altogether when called with invalid parameters. However, if you see a more clean way to intercept/avoid that exception, that could be an alternative as well :) 

I do agree with having a more useful comment describing the nature of the issue in the code. I will add it tomorrow!

ricardoV94 committed:
Add multiple value test for logcdf.
Add more informative comment for Gamma and InverseGamma hack.
Update Release note.

ricardoV94 committed:
Update release note

ricardoV94 committed:
Update docstrings with valid value types

ricardoV94 commented:
Last couple of revisions are complete. I updated the first message to highlight everything that this PR does in one place, but here are the new changes:

1. Add new test in `check_logcdf` to check whether method works with multiple values or, if not, whether it returns an informative standard `TypeError`. 
2. Changed `Beta` and `StudentT` acconding to 1.
3. Updated docstrings as suggested by @AlexAndorra in https://github.com/pymc-devs/pymc3/pull/4387#discussion_r549825529
4. Updated comments as suggested by @domenzain

ricardoV94 committed:
Update RELEASE-NOTES.md

AlexAndorra commented on pymc3/distributions/continuous.py:
```diff
@@ -2478,7 +2507,17 @@ def logcdf(self, value):
         alpha = self.alpha
         beta = self.beta
-        return bound(tt.log(tt.gammainc(alpha, beta * value)), value >= 0, alpha > 0, beta > 0)
+        # To avoid gammainc C-assertion when given invalid values (#4340)
```
What does "gammainc" mean exactly?

domenzain replied:
the regularized lower gamma function (also known as Gamma P)

ricardoV94 replied:
I added a more descriptive comment

AlexAndorra replied:
Thanks @domenzain and @ricardoV94 -- 2021 is starting off pretty well: I already learned something 😉

AlexAndorra commented on pymc3/distributions/continuous.py:
```diff
@@ -2642,11 +2681,16 @@ def logcdf(self, value):
         """
         alpha = self.alpha
         beta = self.beta
+        # To avoid gammaincc C-assertion when given invalid values (#4340)
```
"gammainc" or "gammaincc"?

domenzain replied:
the regularized _upper_ gamma function, also known as gammaincc or gamma Q.

domenzain replied:
The convention is "inc" for incomplete and "c" for "complementary".
See here for the corresponding SciPy function:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammaincc.html

ricardoV94 replied:
@domenzain  is right. The two methods call a different function `gammainc` vs `gammaincc`. Not a typo ;)

AlexAndorra commented on pymc3/tests/test_distributions.py:
```diff
@@ -597,6 +597,35 @@ def check_selfconsistency_discrete_logcdf(
+                err_msg=str(above_domain),
+            )
+
+        # TODO: Test that logcdf wih invalid parameters is always evaluated to -inf
```
Should this comment refer to an issue or PR? It's hard to have the context with the current line

ricardoV94 replied:
Forgot to remove that one!

AlexAndorra commented on pymc3/distributions/continuous.py:
```diff
@@ -1297,20 +1300,32 @@ def logcdf(self, value):
+        if np.ndim(value):
+            raise TypeError(
+                "Beta.logcdf expects a scalar value but received a {}-dimensional object.".format(
+                    np.ndim(value)
```
Any reason you're not using f-strings here?

ricardoV94 replied:
I was not sure if the oldest supported version of pymc3 already includes f-strings. If it does I would be happy to change it, as I also prefer them.

ricardoV94 replied:
Updated it to use f-strings

AlexAndorra commented on pymc3/distributions/continuous.py:
```diff
@@ -1955,20 +1970,34 @@ def logcdf(self, value):
+        if np.ndim(value):
+            raise TypeError(
+                "StudentT.logcdf expects a scalar value but received a {}-dimensional object.".format(
+                    np.ndim(value)
```
Any reason you're not using f-strings here?

ricardoV94 committed:
Add more informative comments and remove TODO

ricardoV94 committed:
Merge remote-tracking branch 'origin/increase_logcdf_coverage' into increase_logcdf_coverage

ricardoV94 committed:
TypeError: format -> f-strings

ricardoV94 commented on pymc3/distributions/discrete.py:
```diff
@@ -1166,9 +1160,7 @@ def logcdf(self, value):
-                "BetaBinomial.logcdf expects a scalar value but received a {}-dimensional object.".format(
-                    np.ndim(value)
-                )
+                f"HyperGeometric.logcdf expects a scalar value but received a {np.ndim(value)}-dimensional object."
```
Previous `TypeError` message referred erroneously to `BetaBinomial`

ricardoV94 committed:
Ignore finite upper limit in Nat domains.
Move new checks to `check_logcdf`.

ricardoV94 committed:
Use `tt.switch` in `DiscreteUniform` for hard boundary (addresses previously failing test in 32bit OS)

ricardoV94 commented:
I had to do 2 small changes:
1. Skip upper bound check for Distributions with Nat domains (they don't have `inf` as the upper bound, but should be ignored anyway)
2. Tweak the logcdf of the discrete distribution.

It is ready for review :)

AlexAndorra reviewed (approved):
All good now, thanks @ricardoV94 🥳
